### PR TITLE
[3265] Remove user related options from messages screen

### PR DESCRIPTION
**Pull request by MarinTolic**
### 🎯 Goal

closes #3625

This PR aims to remove options such as mute and ban/ block from the message options in messages screen.
It also exposes these actions inside the respective Compose & UI Comp `MessageListViewModel`s so that any users building a custom UI with these options can have an easy time accessing them.

#### Compose:

- Deprecated `MessageActions.MuteUser`
- Removed the mute option from `defaultMessageOptionsState()`
- Exposed muting, baning and shadow banning function inside `MessageListViewModel` so users can have an easier time of replacing the lost functionality with custom UI

#### UI Components:

- Deprecated `MessageListViewModel.Event.BlockUser` and replaced it with more explicit banning and shadow banning events
- Introduced a secondary constructor into `MessageListViewStyle` that does not contain blocking and muting parameters and deprecated the primary
- Deprecated blocking and muting related functions and interfaces from `MessageListViewModel`

### 🎨 UI Changes

| Compose Before | Compose After |
| --- | --- |
| remove_message_options_before | remove_message_options_after |

### 🧪 Testing

1. Without applying the patch open Compose
2. Enter a channel and open message options by clicking on a message not sent by you

Expected: there should be no mute option

<details>

<summary>Provide the patch summary here</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/res/layout/fragment_chat.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/res/layout/fragment_chat.xml b/stream-chat-android-ui-components-sample/src/main/res/layout/fragment_chat.xml
--- a/stream-chat-android-ui-components-sample/src/main/res/layout/fragment_chat.xml	(revision 0e6ffbf532bba72a8a588abdb38e8598e265627a)
+++ b/stream-chat-android-ui-components-sample/src/main/res/layout/fragment_chat.xml	(date 1658915825063)
@@ -40,7 +40,7 @@
         app:layout_constraintStart_toStartOf="parent"
         app:layout_constraintTop_toBottomOf="@+id/messagesHeaderView"
         app:streamUiFlagMessageConfirmationEnabled="true"
-        app:streamUiMuteUserEnabled="false"
+        app:streamUiMuteUserEnabled="true"
         app:streamUiPinMessageEnabled="true"
         />
 
Index: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/viewmodel/MessageListViewModelBinding.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/viewmodel/MessageListViewModelBinding.kt b/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/viewmodel/MessageListViewModelBinding.kt
--- a/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/viewmodel/MessageListViewModelBinding.kt	(revision 0e6ffbf532bba72a8a588abdb38e8598e265627a)
+++ b/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/viewmodel/MessageListViewModelBinding.kt	(date 1658915825047)
@@ -77,8 +77,8 @@
     view.setMessageReactionHandler { message, reactionType ->
         onEvent(MessageReaction(message, reactionType, enforceUnique = enforceUniqueReactions))
     }
-    view.setUserMuteHandler { onEvent(MuteUser(it)) }
-    view.setUserUnmuteHandler { onEvent(MessageListViewModel.Event.UnmuteUser(it)) }
+    view.setUserMuteHandler { onEvent(MessageListViewModel.Event.BanUser(it)) }
+    view.setUserUnmuteHandler { onEvent(MessageListViewModel.Event.BanUser(it)) }
     view.setMessageReplyHandler { cid, message -> onEvent(ReplyMessage(cid, message)) }
     view.setAttachmentDownloadHandler { downloadAttachmentCall ->
         PermissionChecker().checkWriteStoragePermissions(view) {
Index: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt
--- a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt	(revision 0e6ffbf532bba72a8a588abdb38e8598e265627a)
+++ b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt	(date 1658915825057)
@@ -904,13 +904,7 @@
      * @param user The user to mute or unmute.
      */
     private fun updateUserMute(user: User) {
-        val isUserMuted = chatClient.globalState.muted.value.any { it.target.id == user.id }
-
-        if (isUserMuted) {
-            unmuteUser(user.id)
-        } else {
-            muteUser(user.id)
-        }
+        banUser(user.id)
     }
 
     /**
Index: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/messageoptions/MessageOptions.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/messageoptions/MessageOptions.kt b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/messageoptions/MessageOptions.kt
--- a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/messageoptions/MessageOptions.kt	(revision 0e6ffbf532bba72a8a588abdb38e8598e265627a)
+++ b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/messageoptions/MessageOptions.kt	(date 1658915825073)
@@ -38,6 +38,7 @@
 import io.getstream.chat.android.common.state.Delete
 import io.getstream.chat.android.common.state.Edit
 import io.getstream.chat.android.common.state.Flag
+import io.getstream.chat.android.common.state.MuteUser
 import io.getstream.chat.android.common.state.Pin
 import io.getstream.chat.android.common.state.Reply
 import io.getstream.chat.android.common.state.Resend
@@ -131,6 +132,7 @@
     val isTextOnlyMessage = selectedMessage.text.isNotEmpty() && selectedMessage.attachments.isEmpty()
     val hasLinks = selectedMessage.attachments.any { it.hasLink() && it.type != ModelType.attach_giphy }
     val isOwnMessage = selectedMessageUserId == currentUser?.id
+    val isUserMuted = currentUser?.mutes?.any { it.target.id == selectedMessageUserId } ?: false
     val isMessageSynced = selectedMessage.syncStatus == SyncStatus.COMPLETED
     val isMessageFailed = selectedMessage.syncStatus == SyncStatus.FAILED_PERMANENTLY
 
@@ -215,6 +217,15 @@
                 titleColor = ChatTheme.colors.errorAccent
             )
         } else null,
+        if (!isOwnMessage) {
+            MessageOptionItemState(
+                title = if (isUserMuted) R.string.stream_compose_unmute_user else R.string.stream_compose_mute_user,
+                iconPainter = painterResource(id = if (isUserMuted) R.drawable.stream_compose_ic_unmute else R.drawable.stream_compose_ic_mute),
+                action = MuteUser(selectedMessage),
+                titleColor = ChatTheme.colors.textHighEmphasis,
+                iconColor = ChatTheme.colors.textLowEmphasis,
+            )
+        } else null
     )
 }
 
```

</details>

Testing is a little bit tedious in this PR.

1. Apply the patch above. You will notice that the mute actions has been restored in Compose and re-enabled in UI Comp. In both Compose and UI comp the actual actions performed when you click on Mute is Ban. It will not unban automatically when you click again because we only have the `User` and not the `Member`
2. Log in as Oleg and go enter the channel `Powerless Radio`. Oleg is the owner here so he can ban/ shadow ban users
3. In the Compose code, replace the patch change with the calls to all of the new functions in order to test out their function: `muteUser()`, `unmuteUser()`, `banUser()`, `unbanUser()`, `shadowBanUser()`, `removeShadowBanFromUser()` 
4. In UI Comp, test the new events by replacing the change in the patch. You should test out: `BanUser`, `UnbanUser`, `ShadowBanUser`, RemoveShadownBanFromUser

#### General
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

gif

### CODE REVIEW

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/messageoptions/MessageOptions.kt [hunk 1]
```diff
@@ -217,15 +215,6 @@ public fun defaultMessageOptionsState(
                 titleColor = ChatTheme.colors.errorAccent
             )
         } else null,
-        if (!isOwnMessage) {
-            MessageOptionItemState(
```
Comment (MarinTolic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/messageoptions/MessageOptions.kt:221]: More difficult to properly deprecate this option in Compose so I opted in for removing it. Users have an easier time of rebuilding this themselves.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt [hunk 2]
```diff
@@ -906,10 +907,119 @@ public class MessageListViewModel(
...
+
+    /**
+     * Mutes the given user inside this channel.
+     *
+     * @param userId The ID of the user to be muted.
+     * @param timeout The period of time for which the user will
+     * be muted, expressed in minutes. A null value signifies that
+     * the user will be muted for an indefinite time.
+     */
+    public fun muteUser(
```
Comment (MarinTolic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt:924]: The idea of exposing these as public is to allow the users to have an easier time of making up for lost functionality should they need it.

File: stream-chat-android-ui-components/src/main/kotlin/com/getstream/sdk/chat/viewmodel/messages/MessageListViewModel.kt [hunk 3]
```diff
@@ -1003,8 +1061,60 @@ public class MessageListViewModel(
...
+        @Deprecated(
+            "Deprecated in order to bring make the action more explicit." +
+                "Use `MessageListViewModel.ShadowBanUser` if you want to keep the same functionality " +
+                "(shadow banning) or `MessageListViewModel.BanUser` if you want to ban a user.",
+            replaceWith = ReplaceWith(
+                "ShadowBanUser",
+                "package com.getstream.sdk.chat.viewmodel.messages.MessageListViewModel.ShadowBanUser"
+            )
+        )
         public data class BlockUser(val user: User, val cid: String) : Event()
```
Comment (MarinTolic) [inline stream-chat-android-ui-components/src/main/kotlin/com/getstream/sdk/chat/viewmodel/messages/MessageListViewModel.kt:1073]: The new ban/ unban, shadow ban/ remove shadow ban `Event`s are more in line with the API and more explicit.
It's implemented in order to give the user more direct control.

File: stream-chat-android-ui-components/src/main/kotlin/com/getstream/sdk/chat/viewmodel/messages/MessageListViewModel.kt [hunk 4]
```diff
@@ -1003,8 +1061,60 @@ public class MessageListViewModel(
...
 
+        /**
+         * When the user bans another user.
+         *
+         * @param user The user to be blocked.
+         * @param reason The reason for banning the user.
+         * @param timeout The time until the user is automatically unbanned, expressed in minutes.
+         * A null value represent an indefinite ban.
+         */
+        public data class BanUser(
```
Comment (MarinTolic) [inline stream-chat-android-ui-components/src/main/kotlin/com/getstream/sdk/chat/viewmodel/messages/MessageListViewModel.kt:1083]: Unlike `Event.BlockUser`, I opet not to ask for CID here.
We have it provided inside `MessageListViewModel` and I think it maybe awkward to expect this VM to be able to block users belonging to a different channel.

I'm not sure why we conceived it differently initially, if you see me missing something here please comment. :smile:

File: DEPRECATIONS.md [hunk 5]
```diff
@@ -4,6 +4,17 @@ This document lists deprecated constructs in the SDK, with their expected time 
...
+| `MessageListView.UserBlockHandler` | 2022.08.02 <br/>5.6.0 | 2022.09.06 ⌛ | 2022.10.04 ⌛ | The option to block the user from `MessageListView` has been deprecated and will be removed. `UserBlockHandler` will be removed with it too. |
+| `MessageListView::setBlockUserEnabled` | 2022.08.02 <br/>5.6.0 | 2022.09.06 ⌛ | 2022.10.04 ⌛ | The option to block the user from `MessageListView` has been deprecated and will be removed. |
+| `MessageListView` attributes | 2022.08.02 <br/>5.6.0 | 2022.09.06 ⌛ | 2022.10.04 ⌛ | The attributes `streamUiMuteOptionIcon`, `streamUiUnmuteOptionIcon`, `streamUiMuteUserEnabled`, `streamUiBlockOptionIcon` and `streamUiBlockUserEnabled` have been deprecated and will be removed. The options to block and mute users will no longer be contained inside `MessageListView` |
+| `MessageListViewStyle` constructor containing params: `muteIcon`, `unmuteIcon`, `muteEnabled`, `blockIcon` and `blockEnabled`  | 2022.08.02 <br/>5.6.0 | 2022.09.06 ⌛ | 2022.10.04 ⌛ | This constructor has been deprecated. Use the constructor that does not contain these parameters. |
```
Comment (adasiewiczr) [inline DEPRECATIONS.md:17]: We should change deprecation version to `5.7.0`

File: stream-chat-android-ui-components/src/main/kotlin/com/getstream/sdk/chat/viewmodel/messages/MessageListViewModel.kt [hunk 6]
```diff
@@ -1003,8 +1061,60 @@ public class MessageListViewModel(
          * @param user The user to be blocked.
          * @param cid The full channel id, i.e. "messaging:123".
          */
+        @Deprecated(
+            "Deprecated in order to bring make the action more explicit." +
```
Comment (TGazica) [inline stream-chat-android-ui-components/src/main/kotlin/com/getstream/sdk/chat/viewmodel/messages/MessageListViewModel.kt:1065]:
```suggestion
            "Deprecated in order to make the action more explicit." +
```

File: DEPRECATIONS.md (hunk 5 above)
Comment (MarinTolic) [reply DEPRECATIONS.md]: Updated, thanks for the heads up :smile:

**Review by TGazica — CHANGES_REQUESTED** [review]

File: stream-chat-android-ui-components/src/main/kotlin/com/getstream/sdk/chat/viewmodel/messages/MessageListViewModel.kt (hunk 6 above)
Comment (MarinTolic) [reply stream-chat-android-ui-components/src/main/kotlin/com/getstream/sdk/chat/viewmodel/messages/MessageListViewModel.kt]: Changed in a commit through AS.

**Review by TGazica — APPROVED** [review]
Looks and works good

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] UI Changes correct (before & after images)
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Review by adasiewiczr — APPROVED** [review]

**Pull request merged**